pow/src/main/java/tech/pegasys/artemis/pow/Eth1MinGenesisTimeBlockFinder.java

```diff
@@ -0,0 +1,280 @@
+              // If exploring downwards and block timestamp < min genesis time,
+              // then previous block must have been the first valid block.
+              firstValidBlock = SafeFuture.completedFuture(previousBlock);
+              break;
+            default:
+              throw new IllegalStateException(
+                  "Unexpected value: "
+                      + calculateCandidateGenesisTimestamp(block.getTimestamp())
+                          .compareTo(Constants.MIN_GENESIS_TIME));
+          }
+          return firstValidBlock;
+        });
+  }
+
+  private SafeFuture<EthBlock.Block> exploreBlocksUpwards(EthBlock.Block previousBlock) {
+    UnsignedLong previousBlockNumber = UnsignedLong.valueOf(previousBlock.getNumber());
+    UnsignedLong newBlockNumber = previousBlockNumber.plus(UnsignedLong.ONE);
+    SafeFuture<EthBlock.Block> blockFuture = eth1Provider.getEth1BlockFuture(newBlockNumber);
+    return blockFuture.thenCompose(
+        block -> {
+          SafeFuture<EthBlock.Block> firstValidBlock;
+          switch (compareBlockTimestampToMinGenesisTime(block)) {
+            case 1:
+            case 0:
+              // If exploring upwards and block timestamp >= min genesis time,
+              // then current block must be the first valid block.
+              firstValidBlock = SafeFuture.completedFuture(block);
+              break;
+            case -1:
+              // If exploring upwards and block timestamp < min genesis time,
+              // then previous block must have been the first valid block.
+              firstValidBlock = exploreBlocksUpwards(block);
+              break;
+            default:
+              throw new IllegalStateException(
+                  "Unexpected value: "
+                      + calculateCandidateGenesisTimestamp(block.getTimestamp())
+                          .compareTo(Constants.MIN_GENESIS_TIME));
+          }
+          return firstValidBlock;
+        });
+  }
+
+  /**
+   * Subscribes to latest block events and sees if the latest canonical block, i.e. the block at
+   * follow distance, is the first valid block.
+   *
+   * @return first valid block
+   */
+  private SafeFuture<EthBlock.Block> waitForFirstValidBlock() {
+    SafeFuture<EthBlock.Block> firstValidBlockFuture = new SafeFuture<>();
+
+    latestBlockDisposable =
+        eth1Provider
+            .getLatestBlockFlowable()
+            .map(EthBlock.Block::getNumber)
+            .map(number -> number.subtract(ETH1_FOLLOW_DISTANCE.bigIntegerValue()))
+            .map(UnsignedLong::valueOf)
+            .subscribe(
+                blockNumber -> onNewBlock(blockNumber, firstValidBlockFuture),
+                firstValidBlockFuture::completeExceptionally);
+
+    return firstValidBlockFuture;
+  }
+
+  private void onNewBlock(
+      UnsignedLong blockNumber, SafeFuture<EthBlock.Block> firstValidBlockFuture) {
+    eth1Provider
+        .getEth1BlockFuture(blockNumber)
+        .thenAccept(
```

Comment:
What happens if this fails?

Reply:
There is an exceptionally compose at the end of the whole findFirstValidBlockAndPublish() so it gets run again. There is a test for it.

Reply:
https://github.com/PegaSysEng/teku/pull/1228/files#diff-b00a71e2a2f564197b1d296f0075e6c4R194

Reply:
Oh nvm, I misread what you wrote. I will make sure this is handled.